aten/src/ATen/native/Resize.h

```diff
@@ -46,4 +48,58 @@ inline TensorImpl* resize_impl_cpu_(
   return self;
 }
 
+static inline void checkInBoundsForStorage(
+    IntList size,
+    IntList stride,
+    ptrdiff_t storageOffset,
+    StorageImpl* new_storage) {
+  ptrdiff_t storage_size = 1;
+  for (size_t dim = 0; dim < size.size(); ++dim) {
+    // NB: (a tensor with arbitrary 0 dims)'s storage can have any numel.
+    if (size[dim] == 0) {
+      return;
+    }
+    storage_size += (size[dim] - 1) * stride[dim];
+  }
+  AT_CHECK(
+      storageOffset + storage_size <= new_storage->numel(),
+      "setStorage: sizes ", size, ", strides ", stride, ","
+      " and storage offset ", storageOffset,
+      " requiring a storage size of ", storage_size + storageOffset,
+      " are out of bounds for storage with numel ", new_storage->numel());
+}
+
+/**
+ * Set self's storage to be new_storage with sizes, strides, and storageOffset.
+ * (size, stride, storageOffset) must be in bounds for the new storage.
+ */
+inline void setStorage(
+    TensorImpl* self,
+    StorageImpl* new_storage,
+    ptrdiff_t storageOffset,
+    IntList size,
+    IntList stride) {
+  AT_ASSERT(new_storage);
+  checkInBoundsForStorage(size, stride, storageOffset, new_storage);
+
+  /* storage */
+  auto* old_storage = self->storage_.unsafeGetStorageImpl();
+  AT_CHECK(old_storage, "Tensor: invalid null storage");
+  if (old_storage != new_storage) {
+    c10::raw::intrusive_ptr::incref(new_storage);
+    THTensor_stealAndSetStoragePtr(self, new_storage);
```

Comment:
I think this code can be written more safely. Here's some context that will help:

1. `THTensor_stealAndSetStoragePtr` purely exists as a BC mechanism for old-style TH calls. Do not feel obligated to use it. And indeed, in this case, using it is inappropriate.
2. `storage_` on `TensorImpl` is currently public. It shouldn't be. One big reason it's public is to support the implementation of `THTensor_stealAndSetStoragePtr` which needs some sort of way of referring to the mutable reference thingy.

FYI! My recommendation would be to do something like add a new method to `TensorImpl` to do this operation directly on the smart pointer, rather than dealing with the raw pointers. This will be much safer, and much cleaner!

Reply:
That sounds reasonable, thank you for the suggestion :)